docs/developers/debugging/compile_time_regressions.md

```diff
@@ -0,0 +1,202 @@
+  starting point. You'll ultimately want to find a culprit release or git
+  commit that changed the compiler code.
+
+* **How much slower did compilation get?**
+
+  Be specific - did it jump from 1 minute to 2 minutes, or 1 minute to 1 hour?
+  Identifying the scale of the regression can help set the priority to
+  investigate it.
+
+* **What is the full compile command?**
+
+  Try to extract the input program and full list of flags passed to the
+  compiler binary so that others can reproduce what you're seeing. Try to
+  distill this as much as possible to using just native tools (no Python or
+  other framework layers).
+
+* **What environment is the compiler running in?**
+
+  Are you using a `Debug` build, or a release build? What operating system and
+  size machine is running the compiler (e.g. Linux developer machine, or a
+  smaller system)?
+
+## Culprit finding and bisecting
+
+If you only have a rough idea of when something changed and want to narrow that
+down to a specific code change, bisecting can help.
+
+### Running `git bisect`
+
+Building the compiler from source and using
+[`git bisect`](https://git-scm.com/docs/git-bisect) will let you pinpoint
+specific commits in IREE, though it typically won't let you step through changes
+in submodules (e.g. MLIR updates in `third_party/llvm-project/`).
+
+__Tip__: [Configure ccache](../developing_iree/ccache.md) if you'll be rebuilding the compiler while bisecting
+
+A manual workflow with `git bisect` looks like this:
+
+```bash
+git bisect start --first-parent
+git bisect good [<rev>]
+git bisect bad [<rev>]
+
+# Read the prompts from the command as it runs
+# At each step, test the compiler:
+#   git submodule update
+#   cmake --build build/ --target iree-compile
+#   ./build/tools/iree-compile <args>
+#       attach Tracy, observe timing, print IR, etc. to determine if fast or slow
+#       if fast, `git bisect good`
+#       if slow, `git bisect bad`
+#   repeat
+```
+
+An automated workflow can use `git bisect run` and a script:
+
+```shell
+# run_bisect.sh
+git submodule update
+cmake --build build/ --target iree-compile
+# Other logic here
+```
+
+```bash
+git bisect start --first-parent
+git bisect good [<rev>]
+git bisect bad [<rev>]
+git bisect run run_bisect.sh
+```
+
+Other sample scripts:
+
+<details><summary>Compile executable sources individually with a timeout:</summary>
+
+```bash
+#!/bin/bash
+
+set -e
+set -o pipefail
+
+# Configure this based on what your thresholds are.
+TIMEOUT_SECONDS_FOR_COMPILING_EACH_SOURCE=10
+
+# Store git version hash, so we can dump artifacts to unique directories later.
+GIT_SHA="$(git rev-parse --short HEAD)"
+
+echo "** Building iree-compile at ${GIT_SHA} **"
+
+# The `git bisect` command only checks out a commit, so update submodules.
+git submodule update
+
+# Build the compiler. You'll want ccache configured to make this fast!
+cmake --build ../iree-build/ --target iree-compile
```

Comment:
Definitely here it should return different error if compilation fails (in theory we never submit broken code ...)

Reply:
Hmmm... https://git-scm.com/docs/git-bisect#_bisect_run

> The special exit code 125 should be used when the current source code cannot be tested. If the script exits with this code, the current revision will be skipped

I don't really want to wrap commands in sample script with error handling, but some patterns could be minimally invasive (https://www.xmodulo.com/catch-handle-errors-bash.html). Would something like this work?
```bash
skip_on_error()
{
  echo "Skipping due to error: $1"
  exit 125  # Special exit code for `git bisect skip`
}
cmake --build ../iree-build/ --target iree-compile || skip_on_error "CMake build failed"
```

@GMNGeoffrey any suggestions?

Reply:
I think that works. Pushed that in one case. I'm keeping the basic `git bisect` sample as simple as possible though.